contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,92 @@
+
+To listen to minion membership changes, master uses an Endpoints informer. An informer update is triggered for every endpoint addition and removal. For each informer update, master sends a `GetMinionName()` call to every endpoint to obtain their identity. If master is aware of the minion, then it updates the minion's `Endpoint`. Otherwise, master adds a new entry to `MinionInfo` and spawns a new goroutine. If a minion no longer exists, master sends a signal through the `StopCh` of the goroutine handling the minion associated with this endpoint, then deletes the associated entry. **TODO**: Master must wait until goroutine finishes before `MinionInfo` entry is removed, but this wait could delay the update of other Endpoints. This is also related to gRPC backoff strategy. Need to handle this.
+
+At initialization, master iterates through fetched endpoints and adds them appropriately to the system as described above. 
+
+It's possible for a new pod to reuse the IP of a deleted pod, so even if the endpoint stays the same over time, the associated minion might be different. To fix this issue, master can pass the target minion's name in every RPC call, and have each minion verify the name. Eventually, minion discovery will assign the Endpoint to the correct minion.
+
+In the case of network partition or minion failure during the `GetMinionStatus()` call, retry a few times, then skip this minion. If the Endpoint is unreachable, the Informer will eventually be triggered, which will correct this scenario.
+
+## Design 2: Minion-Heavy
+
+### High Level Overview
+Another possible design is to leave most of provisioner logic inside minions, only moving PV API interactions to the master. The minion still maintains a PV cache, which is populated by the master, but the difference in this design is it's solely updated by the master. Without this difference, managing the correctness of cache data becomes more difficult, especially when the master fails. When a minion needs to update a PV object, it puts the request in a queue, which is periodically polled by the master.
+
+### Minion Provisioner
+The minion provisioner retains most key components from the previous design, including the Cache, Discoverer, and Deleter. The Populator is moved to master because it interacts with the API server. In order to interact with the master, it exposes the following gRPC calls:
+
+| RPC                      | Description |
+| --- | --- |
+| `Add/Update/RemovePV(PersistentVolume)` | Updates the Cache with the given PV. Called by master when PV informer is triggered. |
+| `PushCurrentPVs([]PersistentVolume)`    | Merges given list of PVs with the Cache. Used by the Populator. |
+| `GetPVRequest()`                        | Peeks at the next unhandled request in the queue. |
```

Comment:
Again, this could return `stream PVRequest` so minion could return new PVs as they are created/mounted without excessive polling